unitest: add flush unittest

LindaYu17 opened the pull request:
Signed-off-by: Linda Yu <linda.yu@intel.com>

xuguangxin committed:
common: add thread abstract for yami

it use executor pattern, user can post and send message to thread.
thread will running until you call stop.

xuguangxin committed:
basedec: add setFormat and ensureProfile to make derived class life easier

Currently, different decoder use different logical to handle format change
and context creating.It have many drawbacks
    1. It’s hard to make sure all decoder have same api behavious.
    2. It duplicate many code in derived class
This patch created a base class function to make derived class’s life easier

xuguangxin committed:
common: add unit test for Thread

xuguangxin committed:
unittest: add decoder api unit test

All decoder should follow this api behaviours.
This patch only add h264, we will add more in future

This unit test will expose following unit test issues.
Mainly because we did not return format change before actually
allocate the surfaces.

    [----------] 1 test from VaapiDecoder/DecodeApiTest
    [ RUN      ] VaapiDecoder/DecodeApiTest.Format_Change/0
    libva info: VA-API version 0.39.3
    libva info: va_getDriverName() returns 0
    libva info: Trying to open /myopt/yami/vaapi/lib/dri/i965_drv_video.so
    libva info: Found init function __vaDriverInit_0_39
    libva info: va_openDriver() returns 0
    DecoderApi_unittest.cpp:335: Failure
    Value of: params->width
      Actual: 16
    Expected: m_width
    Which is: 0
    DecoderApi_unittest.cpp:336: Failure
    Value of: params->height
      Actual: 16
    Expected: m_height
    Which is: 0
    DecoderApi_unittest.cpp:335: Failure
    Value of: params->width
      Actual: 16
    Expected: m_width
    Which is: 8
    DecoderApi_unittest.cpp:336: Failure
    Value of: params->height
      Actual: 32
    Expected: m_height
    Which is: 8
    DecoderApi_unittest.cpp:360: Failure
    Value of: output->crop.width
      Actual: 8
    Expected: format.width
    Which is: 16
    DecoderApi_unittest.cpp:361: Failure
    Value of: output->crop.height
      Actual: 18
    Expected: format.height
    Which is: 16
    DecoderApi_unittest.cpp:414: Failure
    Value of: outFrames
      Actual: 1
    Expected: inFrames
    Which is: 3
    [  FAILED  ] VaapiDecoder/DecodeApiTest.Format_Change/0, where GetParam() = video/h264 (1 ms)

xuguangxin committed:
interface: add fields description for VideoFormatInfo

xuguangxin committed:
h264dec: use ensureSurfacePool and ensureProfile

this change remove duplicate code in h264 and make it follow
same behaviors as other codecs.

fix format change thing, but still have some issues
    [----------] 1 test from VaapiDecoder/DecodeApiTest
    [ RUN      ] VaapiDecoder/DecodeApiTest.Format_Change/0
    libva info: VA-API version 0.39.3
    libva info: va_getDriverName() returns 0
    libva info: Trying to open /myopt/yami/vaapi/lib/dri/i965_drv_video.so
    libva info: Found init function __vaDriverInit_0_39
    libva info: va_openDriver() returns 0
    DecoderApi_unittest.cpp:194: Failure
    Value of: outFrames
      Actual: 3
    Expected: inFrames
    Which is: 4
    DecoderApi_unittest.cpp:149: Failure
    Value of: output->crop.width
      Actual: 8
    Expected: format.width
    Which is: 16
    DecoderApi_unittest.cpp:150: Failure
    Value of: output->crop.height
      Actual: 18
    Expected: format.height
    Which is: 16

xuguangxin committed:
mpeg2dec: use base's setFormat and ensureProfile

use base class's setFormat and ensureProfile can
simplify the flow in mpeg2 decoder, more important,
it will get unified API behavior

 this will break yamidecode, need more investigate about the decoder

xuguangxin committed:
unittest: remove add frame data for decoder test

We only add h264 yet, but we will add more in future

xuguangxin committed:
v4l2base: remove usePixmap since nobody use it

xuguangxin committed:
v4l2dec: delay decoder create to start

this is preparation for next refact

xuguangxin committed:
v4l2: refact, remove platform related display create

xuguangxin committed:
v4l2: refact, remove all android related code

xuguangxin committed:
v4l2: refact, move decoder related to v4l2decoder

xuguangxin committed:
v4l2: add BufferPipe for input/output abstract

xuguangxin committed:
refact: refact v4l2decoder

many improvements in this patch
1. remove extra vpp
2. do not use deprecated api like releaseLock
3. iosolate output related code to V4L2Output
4. use Thread abstact to make decoder more flexible

xuguangxin committed:
decoders: return YAMI_DECODE_NO_SURFACE, if no surface

this is a preparation for no blocking decode in following commits

xuguangxin committed:
decoderpool: do not block caller in surface allocation

block decode thread will make user's life complicate.
many issues related to this
https://github.com/01org/libyami/issues/493
https://github.com/01org/libyami/issues/564
remove it will make our code clear too.
This commit break v4l2 decoder, we will fix it later

xuguangxin committed:
interface: move YAMI_DECODE_NO_SURFACE to no fatal section

xuguangxin committed:
interface; fix typo in SurfaceAllocator

the user should be void* instead of void**

xuguangxin committed:
interface: add getSurface and putSurface to SurfaceAllocParams

This give user a chance to control which surface yami can decode to.

xuguangxin committed:
decoderpool: use getSurface and putSurface api

xuguangxin committed:
v4l2: wrapper, fix typo for __ENABLE_EGL__

xuguangxin committed:
interface: add VIDEO_DATA_MEMORY_TYPE_EXTERNAL_DMA_BUF

LindaYu17 committed:
unitest: add flush unittest

Signed-off-by: Linda Yu <linda.yu@intel.com>

xuguangxin commented on decoder/DecoderApi_unittest.cpp:
```diff
@@ -53,6 +53,20 @@ static FrameData g_h264data[] = {
 };
 
+static FrameData g_h264data_flush[] = {
+    g_avc8x8I,
```
please use g_h264data for flush too. you can add more frames to g_h264data

LindaYu17 replied:
fixed

xuguangxin commented on decoder/DecoderApi_unittest.cpp:
```diff
@@ -70,14 +84,42 @@ class TestDecodeFrames {
+        return false;
+    }
+
+    bool getFrameByIndex(VideoDecodeBuffer& buffer, FrameInfo& info, int32_t index)
```
you can add a seek to start function instead of this

LindaYu17 replied:
fixed

xuguangxin commented on decoder/DecoderApi_unittest.cpp:
```diff
@@ -70,14 +84,42 @@ class TestDecodeFrames {
+        if (d.m_data != NULL) {
+            buffer.data = (uint8_t*)d.m_data;
+            buffer.size = d.m_size;
+            buffer.timeStamp = m_timestamp;
```
please do not put timestap thing in TestDecodeFrames, you can handle this in TEST_P(DecodeApiTest, Flush)

LindaYu17 replied:
fixed

xuguangxin commented on decoder/DecoderApi_unittest.cpp:
```diff
@@ -216,6 +260,68 @@ TEST_P(DecodeApiTest, Format_Change)
+    int32_t size = frames.getFrameCount();
+    int32_t outFrames = 0;
+ 
+    for (int i = 0; i < size + 1; i++) {
```
wired, why size + 1

LindaYu17 replied:
fixed.
it was used to indecate EOS before.

xuguangxin commented on decoder/DecoderApi_unittest.cpp:
```diff
@@ -50,6 +50,11 @@ static FrameData g_h264data[] = {
     g_avc8x16,
     g_avc16x16,
+    g_avc8x8I,
+    g_avc8x8P,
```
Hi @LindaYu17 
Do not change IBP pattern, it will make wrong deocde image, you can dupplicate he original pattern.
IPBIIIPBIPB etc.

LindaYu17 replied:
done

xuguangxin commented on decoder/DecoderApi_unittest.cpp:
```diff
@@ -95,13 +116,15 @@ class TestDecodeFrames {
         const char* mime)
         : m_data(data)
         , m_mime(mime)
+        , m_timestamp(0)
```
Please do not put time stamp here, the time stamp is just for the flush case, you may need put the timestap at flush case only

LindaYu17 replied:
done

LindaYu17 committed:
unittest: add vp8 test data

Signed-off-by: Linda Yu <linda.yu@intel.com>

LindaYu17 committed:
vp8dec: use base's setFormat and ensureProfile

Signed-off-by: Linda Yu <linda.yu@intel.com>

xuguangxin commented on decoder/DecoderApi_unittest.cpp:
```diff
@@ -59,6 +59,13 @@ static FrameData g_h264data[] = {
 };
 
+static FrameData g_vp8data[] = {
+    g_vp8_8x8I,
```
could we add more data like 8x16, 16x16 etc?

xuguangxin replied:
any update on this?

xuguangxin commented on decoder/vaapidecoder_vp8.cpp:
```diff
@@ -83,52 +83,14 @@ YamiStatus VaapiDecoderVP8::ensureContext()
-    m_hasContext = true;
-
-    if (resetContext)
+    if (setFormat(m_frameWidth, m_frameHeight, m_frameWidth, m_frameHeight, 8)) {
```
1. why change this from 3 to 8.
2. could we define a enum or const static value for this?

xuguangxin commented on decoder/vaapidecoder_vp8.cpp:
```diff
@@ -518,14 +480,22 @@ YamiStatus VaapiDecoderVP8::decode(VideoDecodeBuffer* buffer)
 {
     YamiStatus status;
     Vp8ParserResult result;
+    if (!buffer || !buffer->data) {
```
instead of duplicate code here is it possible add function like this
`flush(bool discardOutput)`
and call this flush in` void VaapiDecoderVP8::flush(void)` and here

LindaYu17 replied:
done

xuguangxin commented on decoder/vaapidecoder_vp8.cpp:
```diff
@@ -39,6 +39,8 @@ static const uint8_t keyFrameUVModeProbs[3] = { 142, 114, 183 };
 static const uint8_t nonKeyFrameDefaultYModeProbs[4] = { 112, 86, 140, 37 };
 static const uint8_t nonKeyFrameDefaultUVModeProbs[3] = { 162, 101, 204 };
 
+static const uint32_t surfaceNumVP8 = 3;
```
Please use kSurfaceNumVP8 indicate it's a const.

xuguangxin commented on decoder/vaapidecoder_vp8.cpp:
```diff
@@ -511,21 +471,33 @@ void VaapiDecoderVP8::flush(void)
+void VaapiDecoderVP8::flush(void)
+{
+    DEBUG("VP8: flush()");
+    /*FIXME: should output all surfaces in drain mode*/
```
Please remove this

LindaYu17 committed:
VaapiSurface: compare crop size with original size

we need to compare crop width and height with original surface's width
and height, but not with cropped width and height, because it might be
cropped already before.

Signed-off-by: Linda Yu <linda.yu@intel.com>

LindaYu17 committed:
v4l2_codecbase: initialze all members

Signed-off-by: Linda Yu <linda.yu@intel.com>

xuguangxin commented:
merged in zero copy branch
